File: dfs/new/the_maze.py
class Solution(object):
    def hasPath(self, maze, start, destination):
        return self.helper(maze, destination, start[0], start[1])
        
    
    def helper(self, maze, dest, i, j):
        if [i, j] == dest:
            return True
        if maze[i][j] == 2:
            return False
        up, down, left, right = i, i, j, j
        while up > 0 and maze[up-1][j] != 1:
            up -= 1
        while down < len(maze)-1 and maze[down+1][j] != 1:
            down += 1
        while left > 0 and maze[i][left-1] != 1:
            left -= 1
        while right < len(maze[0])-1 and maze[i][right+1] != 1:
            right += 1
        maze[i][j] = 2
        return self.helper(maze, dest, up, j) or self.helper(maze, dest, down, j) or self.helper(maze, dest, i, left) or self.helper(maze, dest, i, right)


# More Elegant
class Solution2:
    def hasPath(self, maze, start, destination):
        m, n = len(maze), len(maze[0])
        def dfs(x, y, stopped):
            if (x, y) in stopped: return False
            stopped.add((x, y))
            if [x, y] == destination:
                return True
            for i, j in ((-1, 0) , (1, 0), (0, -1), (0, 1)):
                newX, newY = x, y
                while 0 <= newX + i < m and 0 <= newY + j < n and maze[newX + i][newY + j] != 1:
                    newX += i
                    newY += j
                if dfs(newX, newY, stopped):
                    return True
            return False
        return dfs(start[0], start[1], set())

File: dfs/the_maze.rb
# Approach 1: BFS
# @param {Integer[][]} maze
# @param {Integer[]} start
# @param {Integer[]} destination
# @return {Boolean}
def has_path(maze, start, destination)
    q, m, n = [start], maze.size, maze[0].size
    dirs = [[0, 1], [0, -1], [1, 0], [-1, 0]] # [right, left, down, up]

    while !q.empty?
        i, j = q.shift()
        maze[i][j] = 2
        return true if [i, j] == destination

        dirs.each do |x, y|
            r, c = i, j
            
            # Keep going in the same direction till you hit a wall
            while (r + x).between?(0, m - 1) && (c + y).between?(0, n - 1) && maze[r + x][c + y] == 0
                r += x; c += y
            end

            q.push([r, c]) if maze[r][c].zero?
        end
    end

    false
end

# Approach 2: DFS
# @param {Integer[][]} maze
# @param {Integer[]} start
# @param {Integer[]} destination
# @return {Boolean}
def has_path2(maze, start, destination)
    dfs(maze, start, destination)
end

DIRS = [[0, 1], [0, -1], [1, 0], [-1, 0]] # [right, left, down, up]
def dfs(maze, s, d)
    m, n = maze.size, maze[0].size
    i, j = s
    return false if maze[i][j] == 2
    maze[i][j] = 2
    return true if s == d

    DIRS.each do |x, y|
        r, c = s

        # Keep going in the same direction till you hit a wall
        while (r + x).between?(0, m - 1) && (c + y).between?(0, n - 1) && maze[r + x][c + y] != 1
            r += x; c += y
        end

        return true if dfs(maze, [r, c], d)
    end

    false
end

# 490. The Maze
# https://leetcode.com/problems/the-maze/description/

# Apporach 1: BFS, Time: O(m * n), Space: O(m * n)
# Approach 2: DFS, Time: O(m * n), Space: O(m * n)

require 'set'
require 'test/unit'
extend Test::Unit::Assertions

maze = [[0,0,1,0,0],
        [0,0,0,0,0],
        [0,0,0,1,0],
        [1,1,0,1,1],
        [0,0,0,0,0]]
assert_equal(has_path(maze, [0,4], [3,2]), false)
maze = [[0,0,1,0,0],
        [0,0,0,0,0],
        [0,0,0,1,0],
        [1,1,0,1,1],
        [0,0,0,0,0]]
assert_equal(has_path(maze, [0,4], [4,4]), true)
maze = [[0,0,1,0,0],
        [0,0,0,0,0],
        [0,0,0,1,0],
        [1,1,0,1,1],
        [0,0,0,0,0]]
assert_equal(has_path2(maze, [0,4], [3,2]), false)
maze = [[0,0,1,0,0],
        [0,0,0,0,0],
        [0,0,0,1,0],
        [1,1,0,1,1],
        [0,0,0,0,0]]
assert_equal(has_path2(maze, [0,4], [4,4]), true)
